scio-core/src/main/java/com/spotify/scio/transforms/BatchedBaseAsyncLookupDoFn.java

```diff
@@ -0,0 +1,329 @@
+    handleCache(future, cache);
+    // make sure semaphore are released when waiting for futures in finishBundle
+    final FutureType unlockedFuture = handleSemaphore(future);
+
+    futures.put(uuid, handleOutput(unlockedFuture, inputsOnly, uuid));
+    requestCount++;
+  }
+
+  private FutureType handleOutput(FutureType future, List<Input> batchInput, UUID key) {
+    return addCallback(
+        future,
+        response -> {
+          batchResponseFn
+              .apply(response)
+              .forEach(
+                  pair -> {
+                    final Triple<Input, Instant, BoundedWindow> originalInput =
+                        inputsMap.get(pair.getLeft());
+                    final Input input = originalInput.getLeft();
+                    final String inputID = this.idExtractorFn.apply(input);
+                    final Output output = pair.getRight();
+                    results.add(
+                        new Result(
+                            input,
+                            success(output),
+                            key,
+                            originalInput.getMiddle(),
+                            originalInput.getRight()));
+                    this.inputsMap.remove(inputID);
+                  });
+          return null;
+        },
+        throwable -> {
+          batchInput.forEach(
+              input -> {
+                final String inputID = this.idExtractorFn.apply(input);
+                final Triple<Input, Instant, BoundedWindow> originalInput = inputsMap.get(inputID);
+                results.add(
+                    new Result(
+                        input,
+                        failure(throwable),
+                        key,
+                        originalInput.getMiddle(),
+                        originalInput.getRight()));
+                this.inputsMap.remove(inputID);
+              });
+          return null;
+        });
+  }
+
+  private FutureType handleSemaphore(FutureType future) {
+    return addCallback(
+        future,
+        ouput -> {
+          semaphore.release();
+          return null;
+        },
+        throwable -> {
+          semaphore.release();
+          return null;
+        });
+  }
+
+  private FutureType handleCache(FutureType future, Cache<String, Output> cache) {
+    return addCallback(
+        future,
+        response -> {
+          batchResponseFn
+              .apply(response)
+              .forEach(
+                  pair -> {
+                    final String inputID = pair.getLeft();
+                    final Output output = pair.getRight();
+
+                    cache.put(inputID, output);
+                  });
+          return null;
+        },
+        throwable -> {
+          return null;
+        });
+  }
+
+  // Flush pending elements errors and results
+  private void flush(Consumer<Result> outputFn) {
+    Result r = results.poll();
+    while (r != null) {
+      outputFn.accept(r);
+      resultCount++;
```

Comment:
`requestCount` is incremented on batch level but `resultCount` is on the element

Reply:
Nice catch .. fixing

Reply:
Ohh .. forget what i said .. these are not counters and only health check to make sure the number of requests match the number of results .. will just make them match.